server/services/jwt/jwt.service.js

```diff
@@ -0,0 +1,153 @@
+      const newUserTokenData = new userTokenModel({
+        userId: payload.userId,
+        accessTokens: [accessToken],
+        refreshTokens: [refreshToken],
+      });
+      await newUserTokenData.save();
+    }
+
+    return { accessToken, refreshToken };
+  } catch (error) {
+    console.log("Error generating tokens:", error);
+    // Return an error response or rethrow the error based on your application needs
+    return { error: "Token generation failed" };
+  }
+};
+
+/**
+ * @description Method to validate a JWT token
+ * @param {string} token - The JWT token to validate
+ * @returns {Object} Decoded payload if the token is valid, otherwise an error message
+ */
+export const validateToken = (token) => {
+  try {
+    if (!JWT_PWD) {
+      throw new Error("Missing required environment variable: JWT_PWD.");
+    }
+
+    const decoded = jwt.verify(token, JWT_PWD);
+    if (decoded) {
+      return { valid: true, decoded };
+    }
+  } catch (error) {
+    console.log("Error validating token:", error);
+    return { valid: false, error: error.message };
+  }
+};
+
+/**
+ * @description Method to regenerate a new access token using the refresh token
+ * @param {string} refreshToken - The refresh token
+ * @returns {Object} Contains the new generated access token or an error message
+ */
+export const regenerateToken = async (refreshToken) => {
+  // I(manas0410) am using token blacklisting
+
+  try {
+    const decoded = jwt.verify(refreshToken, JWT_PWD);
+
+    const userData = await userTokenModel.findOne({
+      userId: decoded.userId,
+    });
+
+    if (userData && userData?.refreshTokens.includes(refreshToken)) {
+      const newAccessToken = jwt.sign(decoded, JWT_PWD, { expiresIn: "1h" });
+      return { newAccessToken };
+    } else {
+      throw new Error("Invalid token is provided");
+    }
+  } catch (error) {
+    return { error: error.message };
+  }
+};
+
+/**
+ * @description Method to destroy a JWT token
+ * @param {string} token - The JWT token to destroy
+ * @returns {Object} Confirmation message of token destruction
+ *  I(manas0410) am using token blacklisting
+ */
+export const destroyToken = async (token, type) => {
+  try {
+    if (type !== "refreshTokens" && type !== "accessTokens") {
+      throw new Error("Please provide a valid token type.");
+    }
+
+    const { userId } = jwt.verify(token, JWT_PWD);
+
+    if (!userId) {
+      throw new Error("token is not valid");
+    }
+
+    // Update the user document to remove the provided token from the specified type of tokens
+    const result = await userTokenModel.updateOne(
+      { userId },
+      { $pull: { [type]: token } }
+    );
+
+    // Check if the update operation was successful
+    if (result.nModified === 0) {
+      throw new Error("Token not found or user not authorized.");
+    }
+
+    return {
+      message:
+        "Token destruction should be handled on the client side or by maintaining a deny list on the server.",
```

Comment:
@Manas0410 - Can we move all hardcoded messages to a property or constant file

Reply:
@xDAnkit  - Fixed